pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,419 @@
+		return err
+	}
+
+	outputFile, err := afero.TempFile(defaultFS, "", "bundle_*.tar.gz")
+	if err != nil {
+		return fmt.Errorf("error when creating temp file: %w", err)
+	}
+	defer func() {
+		if err = outputFile.Close(); err != nil {
+			klog.ErrorS(err, "Error when closing output tar file")
+		}
+		if err = defaultFS.Remove(outputFile.Name()); err != nil {
+			klog.ErrorS(err, "Error when removing output tar file", "file", outputFile.Name())
+		}
+
+	}()
+	klog.V(2).InfoS("Compressing support bundle collection", "name", supportBundle.Name)
+	if _, err = compress.PackDir(defaultFS, basedir, outputFile); err != nil {
+		return fmt.Errorf("error when packaging support bundle: %w", err)
+	}
+
+	return c.uploadSupportBundle(supportBundle, outputFile)
+}
+
+func (c *SupportBundleController) uploadSupportBundle(supportBundle *cpv1b2.SupportBundleCollection, outputFile afero.File) error {
+	klog.V(2).InfoS("Uploading support bundle collection", "name", supportBundle.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while getting uploader: %v", err)
+	}
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("failed to upload support bundle to file server while setting offset: %v", err)
+	}
+	// fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	parsedURL, err := parseUploadUrl(supportBundle.FileServer.URL)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while parsing upload URL: %v", err)
+	}
+	triesLeft := uploadToFileServerTries
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(uploader, supportBundle.Name, parsedURL, &supportBundle.Authentication, outputFile); uploadErr == nil {
+			return nil
+		}
+		triesLeft--
+		if triesLeft == 0 {
+			return fmt.Errorf("failed to upload support bundle after %d attempts", uploadToFileServerTries)
+		}
+		klog.InfoS("Failed to upload support bundle", "UploadError", uploadErr, "TriesLeft", triesLeft)
+		time.Sleep(uploadToFileServerRetryDelay)
+	}
+	return nil
+}
+
+func parseUploadUrl(uploadUrl string) (*url.URL, error) {
+	parsedURL, err := url.Parse(uploadUrl)
+	if err != nil {
+		parsedURL, err = url.Parse("sftp://" + uploadUrl)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if parsedURL.Scheme != "sftp" {
+		return nil, fmt.Errorf("not sftp protocol")
+	}
+	return parsedURL, nil
+}
+
+func (c *SupportBundleController) uploadToFileServer(up uploader, bundleName string, parsedURL *url.URL, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	joinedPath := path.Join(parsedURL.Path, c.nodeName+"_"+bundleName+".tar.gz")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
```

Comment:
The security configuration you are asking is used when Agent sets up a ssh connection to the target SFTP Server, the current implementation is we do not cache the target server's hostname in Agent's storage, which is thought as a risk by the automatic check. The user who configures the SupportBundleCollection should ensure the sftp server is valid and accessible from Node/ExternalNode. So I don't think it may introduce security risk to Agent.